lazer/contracts/solana/programs/pyth-lazer-solana-contract/src/lib.rs

```diff
@@ -211,3 +268,137 @@ pub struct VerifyMessage<'info> {
     /// This account is not usable with anchor's `Program` account type because it's not executable.
     pub instructions_sysvar: AccountInfo<'info>,
 }
+
+#[derive(Accounts)]
+pub struct VerifyEcdsaMessage<'info> {
+    #[account(mut)]
+    pub payer: Signer<'info>,
+    #[account(
+        seeds = [STORAGE_SEED],
+        bump,
+        has_one = treasury
+    )]
+    pub storage: Account<'info, Storage>,
+    /// CHECK: this account doesn't need additional constraints.
+    pub treasury: AccountInfo<'info>,
+    pub system_program: Program<'info, System>,
+}
+
+fn update_trusted_signer<T: Copy + PartialEq + Default>(
+    stored_num_trusted_signers: &mut u8,
+    stored_trusted_signers: &mut [TrustedSignerInfo<T>],
+    trusted_signer: T,
+    expires_at: i64,
+) -> Result<()> {
+    let num_trusted_signers: usize = (*stored_num_trusted_signers).into();
+    if num_trusted_signers > stored_trusted_signers.len() {
+        return Err(ProgramError::InvalidAccountData.into());
+    }
+    if num_trusted_signers > MAX_NUM_TRUSTED_SIGNERS {
+        return Err(ProgramError::InvalidAccountData.into());
+    }
+    let mut trusted_signers = stored_trusted_signers[..num_trusted_signers].to_vec();
+    if expires_at == 0 {
+        // Delete
+        let pos = trusted_signers
+            .iter()
+            .position(|item| item.pubkey == trusted_signer)
+            .ok_or(ProgramError::InvalidInstructionData)?;
+        trusted_signers.remove(pos);
+    } else if let Some(item) = trusted_signers
+        .iter_mut()
+        .find(|item| item.pubkey == trusted_signer)
+    {
+        // Modify
+        item.expires_at = expires_at;
+    } else {
+        // Add
+        trusted_signers.push(TrustedSignerInfo {
+            pubkey: trusted_signer,
+            expires_at,
+        });
+    }
+
+    if trusted_signers.len() > trusted_signers.len() {
+        return Err(ProgramError::AccountDataTooSmall.into());
+    }
+    if trusted_signers.len() > MAX_NUM_TRUSTED_SIGNERS {
+        return Err(ProgramError::InvalidInstructionData.into());
```

Comment:
I think it's better to have more detailed error messages. I remember when I was debugging an error on mainnet it was really hard to figure out where exactly an error was happening because a single error message can be thrown in multiple places.

Reply:
I agree. Let's handle it as a separate task though.